robusta_krr/core/runner.py

```diff
@@ -22,13 +23,23 @@
 from robusta_krr.utils.progress_bar import ProgressBar
 from robusta_krr.utils.version import get_version
 
+async def gather_with_concurrency(n: int, *coros):
```

Comment:
https://github.com/robusta-dev/krr/pull/82 is not enough. Data race may happen if there are hundreds of deployment.

Reply:
Can you show/explain me when can it fail? Executor can only process max_workers amount of tasks at a time and we were worried exactly about limiting the amount of requests sent to prometheus.

What data race exactly we should be aware of? As I am aware, we do not modify state the way to get a data race.

Reply:
> Can you show/explain me when can it fail? Executor can only process max_workers amount of tasks at a time and we were worried exactly about limiting the amount of requests sent to prometheus.
> 
> What data race exactly we should be aware of? As I am aware, we do not modify state the way to get a data race.

If we have many deployments, let's say 1000. Then 1000 coroutines (aka. tasks) will be created concurrently to gather data.

https://github.com/robusta-dev/krr/blob/9cdc002b9f4a754976dafe5a0452b3ac0242a669/robusta_krr/core/runner.py#L140

Since the max number of workers has been set to a fixed number, for example 6 by default, multiple promQL requests will be queued up. Speically there would be 1000 promQL requested created by `add_history_pods` from different coroutines,

https://github.com/robusta-dev/krr/blob/9cdc002b9f4a754976dafe5a0452b3ac0242a669/robusta_krr/core/integrations/prometheus/metrics_service/prometheus_metrics_service.py#L156

So the second step of a single task cannot be started until all `add_history_pods` are finished. Then all tasks must hold their data in the memory. (I'm not an expert of Python. Pls correct me if I said something wrong)

From my observation, the memory of the process was increasing (> 2GB) while the progress bar was still.

Reply:
Yeah, I think you are right
But let's create a separate issue and I will try to fix it in a separate PR

Reply:
Want to take a deeper look into this one
So please remove the `gather_with_concurrency` for now

Reply:
> Want to take a deeper look into this one So please remove the `gather_with_concurrency` for now

Removed now